src/main/java/org/folio/service/transactions/batch/BatchTransactionService.java

```diff
@@ -0,0 +1,229 @@
+        List<Transaction> transactionsForType = getByType(trType, holder.getAllTransactionsToCreate());
+        if (!transactionsForType.isEmpty()) {
+          getBatchServiceForType(trType).prepareCreatingTransactions(transactionsForType, holder);
+          totalPrepared += transactionsForType.size();
+        }
+      }
+      logger.info("Successfully prepared {} transactions for creation", totalPrepared);
+    } catch (Exception ex) {
+      List<String> ids = holder.getAllTransactionsToCreate().stream().map(Transaction::getId).toList();
+      logger.error("Failed to prepare transactions for creation, ids = {}", ids, ex);
+      throw ex;
+    }
+  }
+
+  private void updatesForUpdatingTransactions(BatchTransactionHolder holder) {
+    if (holder.getAllTransactionsToUpdate().isEmpty()) {
+      return;
+    }
+    try {
+      // NOTE: holder.getAllTransactionsToUpdate() can change after processing payments/pending payments,
+      // so we don't group transactions by type before processing them.
+      int totalPrepared = 0;
+      for (TransactionType trType : transactionTypesInOrder) {
+        List<Transaction> transactionsForType = getByType(trType, holder.getAllTransactionsToUpdate());
+        if (!transactionsForType.isEmpty()) {
+          getBatchServiceForType(trType).prepareUpdatingTransactions(transactionsForType, holder);
+          totalPrepared += transactionsForType.size();
+        }
+      }
+      logger.info("Successfully prepared {} transactions for update", totalPrepared);
+    } catch (Exception ex) {
+      List<String> ids = holder.getAllTransactionsToUpdate().stream().map(Transaction::getId).toList();
+      logger.error("Failed to prepare transactions for update, ids = {}", ids, ex);
+      throw ex;
+    }
+  }
+
+  private void updatesForPatchingTransactions(BatchTransactionHolder holder) {
+    List<TransactionPatch> transactionPatches = holder.getAllTransactionPatches();
+    if (transactionPatches.isEmpty()) {
+      return;
+    }
+    throw new HttpException(500, "transactionPatches: not implemented");
+  }
+
+  private Future<Void> applyChanges(BatchTransactionHolder holder, DBConn conn) {
+    return createTransactions(holder, conn)
+      .compose(v -> updateTransactions(holder, conn))
+      .compose(v -> deleteTransactions(holder, conn))
+      .compose(v -> updateBudgets(holder, conn));
+  }
+
+  private Future<Void> createTransactions(BatchTransactionHolder holder, DBConn conn) {
+    List<Transaction> transactions = holder.getAllTransactionsToCreate();
+    if (transactions.isEmpty()) {
+      return succeededFuture();
+    }
+    return transactionDAO.createTransactions(transactions, conn)
+      .onSuccess(ids -> logger.info("Batch transactions: successfully created {} transactions", transactions.size()))
+      .onFailure(t -> logger.error("Batch transactions: failed to create transactions, transactions = {}",
+        transactions, t))
```

Comment:
Transaction has not toString() implemented, we can use Json.encode() as an option, the same for other methods in this class